### PR TITLE
implement a parallel transform for data preprocessing

### PR DESCRIPTION
**Pull request by QiJune**
Using py_function in `tf.data.Dataset` could not be parallelized due to "GIL" of Python Interpreter. For more details, please refer to this [issue](https://github.com/tensorflow/tensorflow/issues/15694#issuecomment-401857127) and [question](https://stackoverflow.com/questions/48484689/parallelism-isnt-reducing-the-time-in-dataset-map).

The implementation of ParallelTransform takes the [DataLoader](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py) of PyTorch as a reference. It could parallelize the data transform process with low latency.

### CODE REVIEW

File: elasticdl/python/data/parallel_transform.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import itertools
+from multiprocessing import Process, Queue
+
+
+def worker_loop(records, index_q, result_q, transform_fn):
```
Comment (workingloong) [inline elasticdl/python/data/parallel_transform.py:5]: Add a docstring to explain those arguments.

Comment (QiJune) [reply elasticdl/python/data/parallel_transform.py:5]: Done

File: elasticdl/python/data/parallel_transform.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    while True:
+        index = index_q.get()
+        if index is None:
+            break
+        record = records[index]
+        res = transform_fn(record)
+        result_q.put((index, res))
+
+
+class ParallelTransform(object):
```
Comment (brightcoder01) [inline elasticdl/python/data/parallel_transform.py:28]: Can we add an overall description for the ParellelTransform?

Comment (QiJune) [reply elasticdl/python/data/parallel_transform.py:28]: Done

File: elasticdl/python/data/parallel_transform.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+       We transform the data samples one by one. Once we call `next_data`
+       method to get a transformed result, the next index is put to a
+       index queue and triggers a worker process.
+
+    2. The output order is strictly the same with the input order.
+       We use a `cache_dict` to adjust the order, since the result order
+       getting from result queue is sequential sometimes.
+    """
+
+    def __init__(self, records, num_workers, transform_fn):
```
Comment (skydoorkai) [inline elasticdl/python/data/parallel_transform.py]: `num_workers` is a little confusing in ElasticDL.
`num_parallel_processes` is better.

Comment (QiJune) [reply elasticdl/python/data/parallel_transform.py]: Done

**Review by skydoorkai — APPROVED** [review]

**Pull request merged**